doc/icache.rst

```diff
@@ -149,3 +167,53 @@ Cache invalidation
+This is ``instr_pmp_err_i``.
+If the PMP block disallows a fetch for a certain address, it will squash the outgoing memory request entirely and set ``instr_pmp_err_i``.
+If that happens, the cache behaves as if the request was granted but returned some time later with an error.
+Note that it is possible for ``instr_gnt_i`` and ``instr_pmp_err_i`` to be high on the same cycle (if the memory system ties ``instr_gnt_i`` to ``instr_req_o``, for example).
+In that case, the error signal takes precedence.
+
+Fetched instructions are returned to the core using ports ``ready_i``, ``valid_o``, ``rdata_o``, ``addr_o`` and ``err_o``.
+This interface uses ready/valid handshaking in the usual way (a transaction is signalled by ready and valid being high; if ready or valid goes high, it will remain high until the transaction goes through; if valid goes high, the other output signals will remain stable until the transaction goes through).
+The ``rdata_o`` signal contains 32 bits of instruction data, which was fetched from ``addr_o``.
+The ``err_o`` signal will be high if the instruction fetch failed (either with ``instr_pmp_err_i`` or ``instr_err_i``).
+
+Inside the cache is an address counter.
+If ``branch_i`` is asserted then the address counter will be set to ``addr_i`` and the next instruction that is passed to the core will be the one fetched from that address.
+On other cycles, the address counter will be incremented every time an instruction is passed to the core (with a ``ready_i`` / ``valid_o`` transaction).
+This increment depends on the instruction data (visible at ``rdata_o``): if the bottom bits of the instruction are not ``2'b11``, the instruction is considered to be compressed and the address will be incremented by 2.
+If the bottom bits of the instruction are ``2'b11`` then the instruction is considered to be uncompressed and the address will be incremented by 4.
+
+When the cache comes out of reset, it may start fetching memory as soon as ``req_i`` goes high.
+However, the address counter is not initialised on reset.
+This means the behaviour of the I$ is unspecified unless ``branch_i`` is asserted on or before the first cycle that ``req_i`` is asserted after reset.
+If that is not true, there's nothing to stop the cache fetching from random addresses.
+
+The ``req_i`` port signals to the cache that the core is awake and will start requesting instructions soon.
+If ``req_i`` becomes false, the cache may do a few more instruction fetches to fill a cache line, but will stop fetching when that is done.
+The cache will not do any instruction fetches after this until ``req_i`` goes high again.
+The cache assumes that the core will not assert ``ready_i`` when ``req_i`` is low.
+
+The ``icache_enable_i`` signal controls whether the cache stores fetched data.
+If the signal is false, fetched data will be discarded on a branch (or after enough addresses have been fetched in a straight line).
```

Comment:
This might be a bit ambiguous. Buffered data is discarded on a branch only. The cache will cease prefetching when it runs out of resources, but it's not discarding any data at that point.

Reply:
Ok, I'm going to try rejigging this by explicitly talking about a prefetch buffer.